### ISSUE TITLE
The difference between using Python reasoning and C++reasoning

### ISSUE THREAD

**Issue by Bigknowledges** [ultralytics/ultralytics#20660]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

For the same model and picture. I used Python inference: official PT file inference, onnxruntime inference. I used OpenCV:: dnn, onnxruntime, and libtorch for C++inference. At first, I thought the inconsistent results were due to the loss of accuracy caused by the model export settings and transferring from a. pt file to an. onnx file. It wasn't until the Python onnxruntime inference was completed that it was very similar to the official pt inference and the results were acceptable. But when I use C++reasoning, the jagged edges are magnified to an unacceptable size. I have tried many training parameters but have not been able to achieve satisfactory results. But Python language inference does not have this problem, which is strange. I hope to receive some advice (I didn't find this problem before, until this time when my training target was relatively small).

![Image](https://github.com/user-attachments/assets/3fab91d4-0327-4fae-8abf-85740227bedf)

![Image](https://github.com/user-attachments/assets/212d54f9-c1a9-4f64-8d84-b0b39b3370bc)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Bigknowledges, thank you for reaching out and sharing your detailed question with the Ultralytics team! 🚀 This is an automated response to help you get started—an Ultralytics engineer will also review your issue and assist you soon.

We recommend checking out the [Docs](https://docs.ultralytics.com/) for more information, including examples for [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage. Many frequently asked questions are already covered there.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) so we can better understand and debug the issue. For inference differences between Python and C++ backends, sharing:
- The exact export commands and parameters you used
- Sample input images
- Inference scripts/snippets for both Python and C++
- Screenshots or outputs highlighting the differences

...will help us assist you more efficiently.

If your inquiry is about deployment or custom inference, please include as much detail as possible, such as environment, library versions, and any model modifications. Also, review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to ensure optimal training and inference performance.

You're welcome to join the Ultralytics community for more interactive support and discussion:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for live chat
- [Discourse](https://community.ultralytics.com/) for structured discussions
- [Subreddit](https://reddit.com/r/Ultralytics) to share experiences and tips

## Upgrade

To rule out any issues already fixed in recent updates, please upgrade to the latest `ultralytics` package and ensure all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) are satisfied in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO models can be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your patience and for using Ultralytics!

**Comment by Y-T-G** [comment]
Are you using same processing like 

https://github.com/ultralytics/ultralytics/tree/main/examples/YOLOv8-Segmentation-ONNXRuntime-Python

**Comment by Ivan-developer0** [comment]
**Hi @Bigknowledges 👋**

**Thanks for your detailed question!**

You're noticing a difference in inference quality between Python and C++ backends when using the same YOLO model:  
- Python inference with official `.pt` and ONNX Runtime gives acceptable results  
- C++ inference (OpenCV dnn, ONNX Runtime, libtorch) causes jagged edges and worse output  

This can happen due to subtle differences in preprocessing, postprocessing, or backend implementation details between Python and C++ environments.

To help you resolve this:

1. **Ensure identical preprocessing steps** (resizing, normalization, padding) are applied in both Python and C++ inference code. Small differences here can cause large output quality changes.  
2. **Verify consistent postprocessing** (NMS thresholds, confidence thresholds) is used across both implementations.  
3. Share your exact **export commands and parameters**, so we can check if any export setting affects C++ compatibility.  
4. Compare your **inference scripts/snippets** for Python and C++ to spot differences.  
5. Try running the **official ONNX Runtime Python example** from Ultralytics [here](https://github.com/ultralytics/ultralytics/tree/main/examples/YOLOv8-Segmentation-ONNXRuntime-Python) as a reference.  

Also:

- Make sure your C++ environment libraries (OpenCV, ONNX Runtime, libtorch) are up-to-date and match the versions recommended by Ultralytics.  
- Upgrade to the latest `ultralytics` Python package as well:  
 
`  pip install -U ultralytics`

- Test on simple known datasets to isolate whether the issue is dataset or deployment related.

**Please provide your export commands, sample inputs, and inference code snippets for both Python and C++ if you want more tailored assistance!**

**Comment by Bigknowledges** [comment]
> Are you using same processing like
> 
> https://github.com/ultralytics/ultralytics/tree/main/examples/YOLOv8-Segmentation-ONNXRuntime-Python

Yes, I did refer to the code here and the inference results using Python onnxruntime are good. The bad thing is the C++- onnx runtime. Later, I posted the C++inference code.

**Comment by Bigknowledges** [comment]
> **Hi [@Bigknowledges](https://github.com/Bigknowledges) 👋**
> 
> **Thanks for your detailed question!**
> 
> You're noticing a difference in inference quality between Python and C++ backends when using the same YOLO model:
> 
> * Python inference with official `.pt` and ONNX Runtime gives acceptable results
> * C++ inference (OpenCV dnn, ONNX Runtime, libtorch) causes jagged edges and worse output
> 
> This can happen due to subtle differences in preprocessing, postprocessing, or backend implementation details between Python and C++ environments.
> 
> To help you resolve this:
> 
> 1. **Ensure identical preprocessing steps** (resizing, normalization, padding) are applied in both Python and C++ inference code. Small differences here can cause large output quality changes.
> 2. **Verify consistent postprocessing** (NMS thresholds, confidence thresholds) is used across both implementations.
> 3. Share your exact **export commands and parameters**, so we can check if any export setting affects C++ compatibility.
> 4. Compare your **inference scripts/snippets** for Python and C++ to spot differences.
> 5. Try running the **official ONNX Runtime Python example** from Ultralytics [here](https://github.com/ultralytics/ultralytics/tree/main/examples/YOLOv8-Segmentation-ONNXRuntime-Python) as a reference.
> 
> Also:
> 
> * Make sure your C++ environment libraries (OpenCV, ONNX Runtime, libtorch) are up-to-date and match the versions recommended by Ultralytics.
> * Upgrade to the latest `ultralytics` Python package as well:
> 
> ` pip install -U ultralytics`
> 
> * Test on simple known datasets to isolate whether the issue is dataset or deployment related.
> 
> **Please provide your export commands, sample inputs, and inference code snippets for both Python and C++ if you want more tailored assistance!**
Okay, thank you for your reply. Below, I will edit the details of the reasoning. Firstly, the versions of OpenCV, ONNX Runtime, and libtorch are relatively new, opencv481，onnx1.17,libtorch2.7，ultralytics 8.3。 Python onnx inference uses officially recommended code, and the model export settings are also officially recommended (yolo export model=best. pt format=onnx opt not specified). The following is the C++onnx code：

pretreatment：

void utils::letterbox(const cv::Mat &image, cv::Mat &outImage,
                      const cv::Size &newShape = cv::Size(640, 640),
                      const cv::Scalar &color = cv::Scalar(114, 114, 114),
                      bool auto_ = true,
                      bool scaleFill = false,
                      bool scaleUp = true,
                      int stride = 32)
{
    cv::Size shape = image.size();
    float r = std::min((float)newShape.height / (float)shape.height,
                       (float)newShape.width / (float)shape.width);
    if (!scaleUp)
        r = std::min(r, 1.0f);

    float ratio[2]{r, r};
    int newUnpad[2]{(int)std::round((float)shape.width * r),
                    (int)std::round((float)shape.height * r)};

    auto dw = (float)(newShape.width - newUnpad[0]);
    auto dh = (float)(newShape.height - newUnpad[1]);

    if (auto_)
    {
        dw = (float)((int)dw % stride);
        dh = (float)((int)dh % stride);
    }
    else if (scaleFill)
    {
        dw = 0.0f;
        dh = 0.0f;
        newUnpad[0] = newShape.width;
        newUnpad[1] = newShape.height;
        ratio[0] = (float)newShape.width / (float)shape.width;
        ratio[1] = (float)newShape.height / (float)shape.height;
    }

    dw /= 2.0f;
    dh /= 2.0f;

    if (shape.width != newUnpad[0] && shape.height != newUnpad[1])
    {
        cv::resize(image, outImage, cv::Size(newUnpad[0], newUnpad[1]));
    }

    int top = int(std::round(dh - 0.1f));
    int bottom = int(std::round(dh + 0.1f));
    int left = int(std::round(dw - 0.1f));
    int right = int(std::round(dw + 0.1f));
    cv::copyMakeBorder(outImage, outImage, top, bottom, left, right, cv::BORDER_CONSTANT, color);
}


Post processing：
bool Yolov8SegOnnx::OnnxBatchDetect(std::vector<cv::Mat>& srcImgs, std::vector<std::vector<OutputSeg>>& output) {
    vector<Vec4d> params;
    vector<Mat> input_images;
    cv::Size input_size(_netWidth, _netHeight);
    //preprocessing
    Preprocessing(srcImgs, input_images, params); //前处理，主要包含letterbox
    //blobFromImages 返回具有NCHW维度顺序的四维Mat。
    cv::Mat blob = cv::dnn::blobFromImages(input_images, 1 / 255.0, input_size, Scalar(0, 0, 0), true, false);
    

    int64_t input_tensor_length = VectorProduct(_inputTensorShape);
    std::vector<Ort::Value> input_tensors;
    std::vector<Ort::Value> output_tensors;
    input_tensors.push_back(Ort::Value::CreateTensor<float>(_OrtMemoryInfo, (float*)blob.data, input_tensor_length, _inputTensorShape.data(), _inputTensorShape.size()));

    output_tensors = _OrtSession->Run(Ort::RunOptions{ nullptr },
        _inputNodeNames.data(),
        input_tensors.data(),
        _inputNodeNames.size(),
        _outputNodeNames.data(),
        _outputNodeNames.size()
    );

    //post-process

    int net_width = _className.size() + 4 + _segChannels;
    int net_height = net_width;
    float* all_data = output_tensors[0].GetTensorMutableData<float>();
    _outputTensorShape = output_tensors[0].GetTensorTypeAndShapeInfo().GetShape();
    _outputMaskTensorShape = output_tensors[1].GetTensorTypeAndShapeInfo().GetShape();

    vector<int> mask_protos_shape = { 1,(int)_outputMaskTensorShape[1],(int)_outputMaskTensorShape[2],(int)_outputMaskTensorShape[3] };
    int mask_protos_length = VectorProduct(mask_protos_shape);
    int64_t one_output_length = VectorProduct(_outputTensorShape) / _outputTensorShape[0];

    for (int img_index = 0; img_index < srcImgs.size(); ++img_index) {
        Mat output0 = Mat(Size((int)_outputTensorShape[2], (int)_outputTensorShape[1]), CV_32F, all_data).t();  //[bs,116,8400]=>[bs,8400,116]
        all_data += one_output_length;

       
        float* pdata = (float*)output0.data;
        int rows = output0.rows;
        std::vector<int> class_ids;
        std::vector<float> confidences;
        std::vector<cv::Rect> boxes;
        std::vector<vector<float>> picked_proposals;  //output0[:,:, 5 + _className.size():net_width]===> for mask


        for (int r = 0; r < rows; ++r) {    //stride
            cv::Mat scores(1, _className.size(), CV_32F, pdata + 4);
            Point classIdPoint;
            double max_class_socre;
            minMaxLoc(scores, 0, &max_class_socre, 0, &classIdPoint);
            max_class_socre = (float)max_class_socre;

            if (max_class_socre >= _classThreshold) 
            {
                vector<float> temp_proto(pdata + 4 + _className.size(), pdata + net_width);
                picked_proposals.push_back(temp_proto);
                //rect [x,y,w,h]
                float x = (pdata[0] - params[img_index][2]) / params[img_index][0];  //x
                float y = (pdata[1] - params[img_index][3]) / params[img_index][1];  //y
                float w = pdata[2] / params[img_index][0];  //w
                float h = pdata[3] / params[img_index][1];  //h
                int left = MAX(int(x - 0.5 * w + 0.5), 0);
                int top = MAX(int(y - 0.5 * h + 0.5), 0);


                class_ids.push_back(classIdPoint.x);
                confidences.push_back(max_class_socre);
                boxes.push_back(Rect(left, top, int(w + 0.5), int(h + 0.5)));
            }
            pdata += net_width;
        }

        float* raw_pred_data = output_tensors[0].GetTensorMutableData<float>();
        
        std::vector<int64_t> pred_shape = output_tensors[0].GetTensorTypeAndShapeInfo().GetShape();
        int pred_total = pred_shape[1] * pred_shape[2]; 

        vector<int> nms_result;
        cv::dnn::NMSBoxes(boxes, confidences, _classThreshold, _nmsThreshold, nms_result);
        std::vector<vector<float>> temp_mask_proposals;
        cv::Rect holeImgRect(0, 0, srcImgs[img_index].cols, srcImgs[img_index].rows);
        std::vector<OutputSeg> temp_output;
        for (int i = 0; i < nms_result.size(); ++i) {
            int idx = nms_result[i];
            OutputSeg result;
            result.id = class_ids[idx];
            result.confidence = confidences[idx];
            result.box = boxes[idx] & holeImgRect;
            temp_mask_proposals.push_back(picked_proposals[idx]);
            temp_output.push_back(result);
        }
        
        MaskParams mask_params;
        mask_params.params = params[img_index];
        mask_params.srcImgShape = srcImgs[img_index].size();
        Mat mask_protos = Mat(mask_protos_shape, CV_32F, output_tensors[1].GetTensorMutableData<float>() + img_index * mask_protos_length);
        for (int i = 0; i < temp_mask_proposals.size(); ++i) {
            GetMask2(Mat(temp_mask_proposals[i]).t(), mask_protos, temp_output[i], mask_params);
        }
        output.push_back(temp_output);
    }

    if (output.size())
        return true;
    else
        return false;
}

**Comment by wujingweb** [comment]
楼主你怎么解决的，我也遇到这个问题了

**Comment by glenn-jocher** [comment]
Hi @wujingweb — this thread is closed; please open a new issue so we can track it, and include: the full text output of `yolo checks`, your exact ONNX export command, and your C++ mask decode function (the part that does sigmoid, upsampling, cropping, and thresholding), plus a small sample image and text logs. For alignment, mirror the Python path shown in the docs’ mask decoding workflow [process_mask_native reference](https://docs.ultralytics.com/reference/utils/ops/#ultralytics.utils.ops.process_mask_native) and resizing back to the original image via [scale_image](https://docs.ultralytics.com/reference/utils/ops/#ultralytics.utils.ops.scale_image).

```bash
yolo checks
```

```python
from ultralytics.utils.checks import collect_system_info
print(collect_system_info())
```
